### DISCUSSION TITLE
can i assign a template by prefix of a new note's name?

### DISCUSSION THREAD

**Discussion by webrunneronSun** [SilentVoid13/Templater#1055] [Help]
Hi 
if this already been asked, please forgive me ask  again. I searched the forum, but did not find the answer.

what i want is: while i press ctrl+N to create a new note, when prompt to input the notes name, by input such as "h-something to create", it will call a template based on the prefix "h-" to apply, and then it re-name the note to "something to create".

thanks!

**Comment by Zachatoo** [comment]
We can use folder templates and a global template that will run for all files in your vault, and all it will do is prompt for the file name and apply a template based on the prefix. Here's an example template (I do something similar [here](https://zachyoung.dev/blog/folder-templates-with-quick-switcher#picker-template), however I use a suggester instead):

```js
<%*
// Fallback template if no or invalid prefix
const DEFAULT_TEMPLATE = 'default-template';
let template = DEFAULT_TEMPLATE;

// List out prefixes and associated templates
// For example, the "p-" prefix uses the "person-template" template
const types = [
  { prefix: 'p', template: 'person-template' },
  { prefix: 't', template: 'term-template' },
];

const title = await tp.system.prompt("Title");
if (title && title.includes('-')) {
  // Grab prefix from prompt
  const [prefix, ...rest] = title.split('-');

  // If prefix matches list of prefixes above, then use the associated template for that prefix
  const type = types.find(x => x.prefix === prefix);
  if (type) {
    template = type.template;
  }

  // Rename file without the prefix
  const newTitle = rest.join('-');
  await tp.file.rename(newTitle);
}

// Apply template
tR += await tp.file.include(`[[${template}]]`)
-%>
```

Here's the steps to setup folder templates:

1. Turn on "Trigger templater on new file creation"
2. Turn on "Enable folder templates"
3. Set the folder to `/` and the template to the template created above

**Reply by Zachatoo** [reply]
Note that "Show tab title bar" must be turned on in Appearance settings for this to work. If that setting is turned off, then you will be prompted _twice_ for a title, which is super annoying.

**Comment by webrunneronSun** [comment]
Thanks @Zachatoo 

That is great!  I will try these soon as I have had been busy with some urgent matters for days and haven't time to check my question. I may need more helps to get it works as I know very little about programming language.

anyway, i can read it and understood it with the explanations, hope I can have it work in my Obsidian vault.

thanks again!

**Comment by webrunneronSun** [comment]
hello @Zachatoo 

I tried the template, when use prefix to create new notes , it works well. but when without prefix, it creates a notes named untitled, but not what I put input.  if i press ctrl+N, it throws an error and creates another notes named untitled1...

 
![image](https://user-images.githubusercontent.com/119090113/227595267-af091a7c-44f8-462e-9e99-b480d36ed005.png)


please help again. thanks.

**Reply by Zachatoo** [reply]
I have updated my original answer to account for this. This is the line I changed.

```js
if (title) {
// Has been changed to
if (title && title.includes('-')) {
```

Basically just checks to see if there is a prefix before trying to do any prefix logic.

**Comment by webrunneronSun** [comment]
thank you very much @Zachatoo  

finally I got it works! thanks again.

but I got another problems with the <% tp.file.title %> which dose not work in my template! please help again.

here is the problem:
![image](https://user-images.githubusercontent.com/119090113/227768512-eda3c277-8759-4f98-aa9a-7e4d7e6978ae.png)

What I did: 

I added some changes to the first snippets you posted. Here's what I added:
 
```
 // Without prefix in title, fall back to default-template
  else {
    await tp.file.rename(title);
  }
``` 
it works now. 

But the <% tp.file.title %> seems not work in the other called templates such as in "default-template.md", "person-template.md" and "term-template.md"

Templates used:

1. default-template.md ( the <% tp.date.now() %> works well, while <% tp.file.title %> doesn't )
```
---
uuid: <% tp.date.now() %>
---
# <% tp.file.title %>

this is default template
this is file name: <% tp.file.title %>
```
2. person-template.md
```
---
uuid: <% tp.date.now() %>
---

this is person template
this is file name <% tp.file.title %>

```

3. term-template.md
```
this is term template
this is file name <% tp.file.(title) %>
```



Snippet after changes made:
```
<<%*
// Fallback template if no or invalid prefix
const DEFAULT_TEMPLATE = 'default-template';
let template = DEFAULT_TEMPLATE;

// List out prefixes and associated templates
// For example, the "p-" prefix uses the "person-template" template
const types = [
  { prefix: 'p', template: 'person-template' },
  { prefix: 't', template: 'term-template' },
];

const title = await tp.system.prompt("Title");
if (title) {
  // Grab prefix from prompt
  const [prefix, ...rest] = title.split('-');

  // If prefix matches list of prefixes above, then use the associated template for that prefix
  const type = types.find(x => x.prefix === prefix);
  if (type) {
    template = type.template;
    // Rename file without the prefix
  const newTitle = rest.join('-');
  await tp.file.rename(newTitle);}
  // Without prefix in title, fall back to default-template
  else {
    await tp.file.rename(title);
  }
}

// Apply template
tR += await tp.file.include(`[[${template}]]`)
-%>

```

**Comment by webrunneronSun** [comment]
hi @Zachatoo 

I figured it out. for the "<" problem was I made a type error, I added a "<" at the very beginning of the template. 

for the "untitled" problem, I use <% tp.config.target_file.basename %>, which I searched and found on the forum, and guess it could be work. then I tested it, it works well, but I still do not understand why it works.

Here are the templates I am using, for people who may need it.

1. template to auto-select template based on prefix, such as p-, or t- or none

```
<%*
// Fallback template if no or invalid prefix
const DEFAULT_TEMPLATE = 'default-template';
let template = DEFAULT_TEMPLATE;

// List out prefixes and associated templates
// For example, the "p-" prefix uses the "person-template" template
const types = [
  { prefix: 'p', template: 'person-template' },
  { prefix: 't', template: 'term-template' },
];

const title = await tp.system.prompt("Title");
if (title) {
  // Grab prefix from prompt
  const [prefix, ...rest] = title.split('-');

  // If prefix matches list of prefixes above, then use the associated template for that prefix
  const type = types.find(x => x.prefix === prefix);
  if (type) {
    template = type.template;
    // Rename file without the prefix
  const newTitle = rest.join('-');
  tR += await tp.file.rename(newTitle);
    }
  // Without prefix in title, fall back to default-template
  else {
  tR += await tp.file.rename(title);
  }
}

// Apply template
tR += await tp.file.include(`[[${template}]]`)
-%>
```



3. if no prefix, template to be used

```
---
uuid: <% tp.date.now() %>
---
# <% tp.config.target_file.basename %>

this is default template
this is file name:  <% tp.config.target_file.basename %>

```

4. if p- is the prefix, template to be used
```
---
tags: person
uuid: <% tp.date.now() %>
---

# <% tp.config.target_file.basename %>
this is person template
this is file name # <% tp.config.target_file.basename %>
```
5. if t- is the prefix, template to be used  
```
---
tags: term
uuid: <% tp.date.now() %>
---


# <% tp.config.target_file.basename %>

this is term template.
this is file name <% tp.config.target_file.basename %>

```
these templates could not be reasonable. but it works. 
So glad that with your helps, I made it works!

thanks again.
